Core: Migrate from `qs` to `picoquery`

43081j opened the pull request:
Moves the `manager-api`, `preview-api` and `ui/manager` to use `picoquery` instead of `qs` - a much smaller and faster alternative.

Note that we still have `qs` in our overall monorepo tree, but that can be tackled in a follow-up PR.

## Checklist for Contributors

### Testing

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests (existing tests should suffice)
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

Existing stories and tests should continue to function correctly.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

43081j committed:
feat: migrate from qs

Moves the `manager-api`, `preview-api` and `ui/amanger` to use
`picoquery` instead of `qs` - a much smaller and faster alternative.

Note that we still have `qs` in our overall monorepo tree, but that can
be tackled in a follow-up PR.

shilman committed:
Use destructured imports

shilman committed:
qs => picoquery

shilman commented:
Looks like there are a few extra qs references that are causing linting to fail!

shilman committed:
Merge branch 'next' into pq

43081j commented:
whoops my mistake! i forgot to run `--task check` i think

looks like you got it building though. let me know if you need any changes or help

shilman commented:
I swapped one of them over, but didn't go the extra distance to do the other two, which are a little more involved 🙈

43081j commented:
no worries, its probably the one i left out on purpose

happy to tackle that in its own PR as long as the rest of the packages build ok

sure means we temporarily have both in the dependency tree but its small enough we'll be ok i think, and i can probably sort it fairly soon

43081j committed:
feat: use js syntax in query strings

Uses the `foo.bar[0]` syntax picoquery provides, rather than dot-syntax
(`foo.bar.0`).

43081j commented:
note there's still one usage left i need to tackle

i have failing tests locally for it but will get to it at some point 👍

43081j committed:
feat (preview-api): move to picoquery

Moves the preview api to use picoquery.

Note that tests will not pass yet.

ghengeveld commented on code/lib/preview-api/src/modules/preview-web/UrlStore.ts:
```diff
@@ -20,19 +20,20 @@ const getQueryString = ({
-  });
-  return qs.stringify(
-    {
+  const search =
+    typeof document !== 'undefined' && document.location.search
+      ? document.location.search.slice(1)
+      : '';
```
I think this would be cleaner:
```suggestion
  const search = document?.location.search.slice(1) || '';
```
`slice(1)` on an empty string still yields an empty string.

ghengeveld commented on code/lib/preview-api/src/modules/preview-web/UrlStore.ts:
```diff
@@ -20,19 +20,20 @@ const getQueryString = ({
+  const { path, selectedKind, selectedStory, ...rest } = parse(search);
+  return (
+    '?' +
+    stringify({
```
Would be nicer to extract the stringify result to a variable, then concatenate using a template string.

ghengeveld commented on code/lib/preview-api/src/modules/preview-web/UrlStore.ts:
```diff
@@ -58,15 +59,18 @@ const getFirstString = (v: ValueOf<qs.ParsedQs>): string | void => {
 export const getSelectionSpecifierFromPath: () => SelectionSpecifier | null = () => {
   if (typeof document !== 'undefined') {
-    const query = qs.parse(document.location.search, { ignoreQueryPrefix: true });
+    const queryStr = document.location.search ? document.location.search.slice(1) : '';
+    const query = parse(queryStr);
```
```suggestion
    const query = parse(document.location.search.slice(1));
```
There's no need to check for `search` to exist. Afaik it'll always exist on `location` and always be a string, even if empty. Slicing an empty string will yield an empty string, so that's safe too.

ghengeveld commented on code/lib/router/src/utils.test.ts:
```diff
@@ -137,12 +137,12 @@ describe('buildArgsParam', () => {
   it('builds arrays', () => {
     const param = buildArgsParam({}, { arr: ['1', '2', '3'] });
-    expect(param).toEqual('arr[0]:1;arr[1]:2;arr[2]:3');
+    expect(param).toEqual('arr%5B0%5D:1;arr%5B1%5D:2;arr%5B2%5D:3');
```
Let's not do that. This would be a breaking change and it looks extremely ugly. I know it's not standards compliant to keep unencoded things in the URL, but it works fine in all browsers and these are only parsed on the client-side (no server involved).

ghengeveld commented on code/lib/router/src/utils.ts:
```diff
@@ -149,11 +153,14 @@ interface Query {
-export const queryFromLocation = (location: Partial<Location>) => queryFromString(location.search);
-export const stringifyQuery = (query: Query) =>
-  qs.stringify(query, { addQueryPrefix: true, encode: false });
+export const queryFromLocation = (location: Partial<Location>) =>
+  queryFromString(location.search ? location.search.slice(1) : '');
```
```suggestion
export const queryFromLocation = (location: Partial<Location>) =>
  queryFromString(location.search.slice(1));
```

ghengeveld commented on code/ui/manager/src/components/preview/utils/stringifyQueryParams.tsx:
```diff
@@ -1,4 +1,4 @@
+import { stringify } from 'picoquery';
 
 export const stringifyQueryParams = (queryParams: Record<string, string>) =>
-  qs.stringify(queryParams, { addQueryPrefix: true, encode: false }).replace(/^\?/, '&');
```
I think we need some unit tests to verify `stringifyQueryParams` doesn't change between the two implementations.

ghengeveld reviewed (dismissed):
I'm happy to move from qs to picoquery, but only if it involves no functional changes. Unfortunately, it seems picoquery urlencodes args, which aren't currently encoded. This would be a breaking change and make the URLs look really ugly, so I'm opposed to changing that behavior.

43081j commented:
> I'm happy to move from qs to picoquery, but only if it involves no functional changes. Unfortunately, it seems picoquery urlencodes args, which aren't currently encoded. This would be a breaking change and make the URLs look really ugly, so I'm opposed to changing that behavior.

this is true, however it is doing the right thing here while qs/storybook hasn't been

the URI spec defines it this way, and is why `URLSearchParams` also encodes them

so we're blocked here i guess. i understand the URLs looked nicer, but they're non-standard and its unlikely pq will ever support that (since it is trying to conform to the standard and be interchangeable with `URLSearchParams` in those cases)

ghengeveld commented:
> alternatively, we could have code inside storybook which literally does `replace('%3B', '[')` etc. i'm not a fan but at least it doesn't result in veering away from standards in PQ and can live here instead

That works for me. It's hacky but at least it's consistent. If being noncompliant turns out to be an issue then we can always change it in a major version release (in which case I'd opt for a different/custom array encoding to avoid `%5B` / `%5D`).

43081j commented:
no worries, that works for me too

we already do some transforms on the parsed and stringified queries so it won't be much to add to that

43081j committed:
Merge branch 'next' into pq

43081j committed:
fix: decode some known query string chars

Decodes certain chars we want to keep in our URIs (e.g. `[`).

43081j commented:
we have some failures ill have to get to at some point another day 👍 

basically what to do about invalid keys since pq will deal with them differently than qs did (for perf mostly)

e.g. `a[b:val` will likely result in `{a: []}`

ghengeveld commented:
> basically what to do about invalid keys since pq will deal with them differently than qs did (for perf mostly)
> 
> e.g. `a[b:val` will likely result in `{a: []}`

That’s okay. Invalid input doesn’t need to be parsed in any particular way. The moody important thing is that data must deserialize back to what it was before serialization, to the extent that that’s possible.

ndelangen committed:
Merge branch 'next' into pq

43081j commented:
i have a busy few days ahead and some travel coming up but ill get to this when i can 👍 

will ping you when i manage to fix the remaining tests

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

43081j commented:
just picking this up again @ghengeveld , any idea what this kind of stuff was trying to catch?

```ts
      expect(parseArgsParam('obj.foo[][a!b]:val;obj.foo.bar:val;obj.baz:val')).toStrictEqual({});
```

this syntax isn't valid anywhere really, `obj.foo[][a!b]`. qs does what it can with it, but its meant to be `obj[foo][][a!b]` or `obj.foo[].a!b`, etc. these inputs are pretty inconsistent, mixing dot-syntax and bracket-syntax

ultimately, its trying to test that nested keys are checked for validity. so maybe i can just update these to be more sensible inputs?

43081j committed:
test: fix parseArgsParam tests

43081j committed:
Merge branch 'next' into pq

43081j committed:
test: add prefix to test queries

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

ndelangen committed:
dedupe

43081j committed:
Merge branch 'next' into pq

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

43081j committed:
Merge branch 'next' into pq

43081j committed:
Merge branch 'next' into pq

JReinhold committed:
Merge branch 'next' into pq

JReinhold commented on code/core/src/router/utils.ts:
```diff
@@ -155,13 +163,33 @@ const encodeSpecialValues = (value: unknown): any => {
+// URIs.
+const decodeKnownQueryChar = (chr: string) => {
+  switch (chr) {
+    case '%20':
+      return '+';
```
Either way, I don't think it matters right now, because when trying out a Date-control on `next`, a Date arg always turns into the UNIX time number , and never the special value `!date(ISO...)`.

It's probably because of these lines, which have been here since the very beginning: https://github.com/storybookjs/storybook/blob/next/code/lib/blocks/src/controls/Date.tsx/#L93-L96

I'm assuming this is a bug still @ghengeveld?

43081j replied:
yeah I suspect if we decode `+`, we lose the knowledge of if it was a space or a plus. but maybe that loss is ok since we treat both as a space? 🤔 

I'm not really sure of the best thing to do here. you will know better than I do, so I'm happy to go with whatever you and the others decide

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

ndelangen commented on code/core/src/preview-api/modules/preview-web/parseArgsParam.test.ts:
```diff
@@ -58,7 +58,7 @@ describe('parseArgsParam', () => {
 
   it('parses Date with timezone offset', () => {
-    const args = parseArgsParam('key:!date(2001-02-03T04:05:06.789+09:00)');
+    const args = parseArgsParam('key:!date(2001-02-03T04:05:06.789%2B09:00)');
```
Are all these updated tests not indications that this is backwards incompatible, and we should delay merging this until the next major version @JReinhold ?

ghengeveld replied:
I don't agree with this updated format to begin with. `+` is valid in a URL. Moreover, this is fully client-side URL parsing so we don't even have to adhere to the specs necessarily.

JReinhold replied:
this instance is what I'm commenting on here:
https://github.com/storybookjs/storybook/pull/28315/files#r1763000236

as for the other tests, it looks to me that they were testing that invalid paths were still "working", and it's fine to me that they don't work the same. I don't see how they could possibly be constructed by us, so the user must have manually written an invalid query param, and I don't feel the need to support that at all.

43081j replied:
elsewhere, we translate spaces to `+` so maybe I also missed a place where we need to do the same

ndelangen committed:
rename variable to indicate we're not using `qs` anymore

ndelangen committed:
dedupe

JReinhold committed:
Discard changes to code/core/src/preview-api/modules/preview-web/parseArgsParam.test.ts

JReinhold committed:
fix parseArgsParams and tests

JReinhold committed:
format

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

ndelangen committed:
fix

ndelangen committed:
Fixes by not re-using query-options meant for args&globals use only

ndelangen committed:
make our color examples lowercase always & fix a bug where we didn't detect uppercase color prefixes

ndelangen committed:
fix

ndelangen committed:
Merge branch 'next' into pr/43081j/28315

JReinhold reviewed (approved):
🎉

ndelangen committed:
Merge branch 'next' into pq

43081j commented:
Great work landing this 🙌

If you run into anything in the pre release, let me know and I'll try help out

JReinhold commented:
Thanks for the help @43081j ❤️
